modopt/opt/algorithms.py

```diff
@@ -78,50 +80,123 @@ class SetUp(Observable):
         Generic step size parameter to override default algorithm
         parameter name (`e.g.` `step_size` will override the value set for
         `beta_param` in `ForwardBackward`)
+    use_gpu : bool, optional
+        Option to use available GPU
 
     """
 
-    def __init__(self, metric_call_period=5, metrics={}, verbose=False,
-                 progress=True, step_size=None, **dummy_kwargs):
+    def __init__(
+        self,
+        metric_call_period=5,
+        metrics=None,
+        verbose=False,
+        progress=True,
+        step_size=None,
+        use_gpu=False,
+        **dummy_kwargs,
+    ):
 
         self.converge = False
         self.verbose = verbose
         self.progress = progress
         self.metrics = metrics
         self.step_size = step_size
-        self._op_parents = ('GradParent', 'ProximityParent', 'LinearParent',
-                            'costObj')
+        self._op_parents = (
+            'GradParent',
+            'ProximityParent',
+            'LinearParent',
+            'costObj',
+        )
 
         self.metric_call_period = metric_call_period
 
         # Declaration of observers for metrics
-        Observable.__init__(self, ["cv_metrics"])
+        super().__init__(['cv_metrics'])
+
+        for name, dic in self.metrics.items():
+            observer = MetricObserver(
+                name,
+                dic['metric'],
+                dic['mapping'],
+                dic['cst_kwargs'],
+                dic['early_stopping'],
+            )
+            self.add_observer('cv_metrics', observer)
+
+        # Check for GPU
+        if use_gpu:
```

Comment:
It seems weird to me that some GPU functionalities are introduced in this PR which should only deal with style right?

Reply:
Hi @zaccharieramzi, @chaithyagr pushed to include the changes he has implemented in Incorporated #97. Since these don't (or at least as far I could see) impact the standard CPU implementation or API I figured it should be OK for now.

I will need to open an issue to standardise this throughout the code, but I think this could also be done for the mid 2021 release.

If there is anything specific that you are concerned about please point it out. I would rather delay the release by a couple of days to sort it out rather than rushing through something that could cause problems later on.